neovimUtils.wrapNeovimUnstable: support vim.o.exrc

BirdeeHub opened the pull request:
wrap neovim config using VIMINIT rather than -u

-u skips several initialization steps

from `:h startup`

```
     a. The system vimrc file is read for initializations.  If
	nvim/sysinit.vim file exists in one of $XDG_CONFIG_DIRS, it will be
	used.  Otherwise the system vimrc file is used. The path of this file
	is given by the |:version| command.  Usually it's "$VIM/sysinit.vim".
```

```
     c. If the 'exrc' option is on (which is NOT the default), the current
	directory is searched for the following files, in order of precedence:
	- ".nvim.lua"
	- ".nvimrc"
	- ".exrc"
	The first that exists is used, the others are ignored.
```

b is ommitted here because skipping b is actually the intent of wrapRc

Using VIMINIT causes the config init.lua to be sourced during step b as intended, bringing it in line with normal config initialization.

With -u, VIMINIT is skipped and never checked anyway.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

BirdeeHub committed:
neovimUtils.wrapNeovimUnstable: support vim.o.exrc

mrcjkb commented on pkgs/applications/editors/neovim/wrapper.nix:
```diff
@@ -193,8 +193,9 @@ let
           ++ lib.optionals finalAttrs.wrapRc [
-            "--add-flags"
-            "-u ${writeText "init.lua" rcContent}"
+            "--set"
+            "VIMINIT"
+            "lua dofile([[${writeText "init.lua" rcContent}]])"
```
question: The general idea of using `VIMINIT` instead of `-u` seems reasonable to me.
But what if the user has set `VIMINIT`? This implementation will ignore that and use the generated init.lua instead. Perhaps the wrapper should only set `VIMINIT` if it isn't already set?

BirdeeHub replied:
> With -u, VIMINIT is skipped and never checked anyway.

^

The problem with -u is it skips ALL that stuff.

BirdeeHub replied:
for what its worth, nixCats does this currently

https://github.com/BirdeeHub/nixCats-nvim/blob/0bb051b15e2c2bd4d63532317a84ca24740c0d1c/builder/wrapper.nix#L121

BirdeeHub replied:
also, @PerchunPak what do you think of the new hosting updates? [:h nixCats.flake.outputs.settings.hosts](https://nixcats.org/nixCats_format.html#nixCats.flake.outputs.settings.hosts)

You can bundle anything as a host now and define categories for it and stuff :)

BirdeeHub replied:
should we add os.getenv?

Check viminit first?

BirdeeHub replied:
hmmm. I would need to check the env var in vimscript actually would probably be better

BirdeeHub replied:
I started writing it.

I dont like the look of `if VIMINIT env var is a string, eval the code in it` in my PRs necessarily XD

BirdeeHub replied:
if they set it in the wrapper it will overwrite.

Do we want it to work at runtime or not is the question.

BirdeeHub replied:
I would argue, if someone wants to redefine viminit at runtime and lose all the generated config, are they best served by using wrapRc = true?

BirdeeHub replied:
`lua local viminit = os.getenv('VIMINIT'); if viminit and viminit ~= "" then vim.cmd(viminit) else dofile([[${writeText "init.lua" rcContent}]])`

BirdeeHub replied:
so, fixing that would need to be done somewhere else.

BirdeeHub replied:
is there a `--set-if-unset` wrapper arg?

BirdeeHub commented on pkgs/applications/editors/neovim/wrapper.nix:
```diff
@@ -193,8 +193,9 @@ let
           ++ lib.optionals finalAttrs.wrapRc [
-            "--add-flags"
-            "-u ${writeText "init.lua" rcContent}"
+            "--set"
```
```suggestion
            "--set-default"
```

BirdeeHub replied:
@mrcjkb yes or no should it be set-default or just --set

BirdeeHub replied:
Actually, i think --set-default is better tbh

mrcjkb reviewed (approved):
I'm on my mobile, so I can't run nixpkgs-review, but this looks like an improvement to me.

PerchunPak commented:
> I'm on my mobile, so I can't run nixpkgs-review, but this looks like an improvement to me.

Instead of running nixpkgs-review, someone could just rebuild their config using this PR

(btw @BirdeeHub the feature is dope)

MattSturgeon commented on pkgs/applications/editors/neovim/wrapper.nix:
```diff
@@ -193,8 +193,9 @@ let
             "${placeholder "out"}/rplugin.vim"
           ]
           ++ lib.optionals finalAttrs.wrapRc [
-            "--add-flags"
-            "-u ${writeText "init.lua" rcContent}"
```
Is there any documentation that needs updating? E.g. anything that mentions the wrapper using `-u`?

I think this should also be mentioned in release notes.

BirdeeHub replied:
I have no idea actually, I can go check.

Id be pretty surprised if it specifically mentions -u?

BirdeeHub replied:
There was, and it was updated now, although the doc check is now showing up in the list of checks, and its failing for Typst and not the changes I made to the nvim section

MattSturgeon reviewed (commented):
So, to oversimplify: the main change here is that `/etc/nvim/sysinit.vim` will be sourced before the wrapped config.[^1]

That's a breaking change, since some users may be relying on system config being ignored.

IMO neither approach is “better,” and either may be more suitable depending on the situation and the user's intention.

I could be persuaded either way regarding which approach is better for _most_ users and should be used by the wrapper; or if both techniques were supported, which should be the default behaviour.

[^1]: and support for the `exrc` option, but I _was_ oversimplifying

BirdeeHub commented:
> (btw @BirdeeHub the feature is dope)

The remote host rework? Yeah ☺️ thank you

BirdeeHub commented:
> So, to oversimplify: the main change here is that `/etc/nvim/sysinit.vim` will be sourced before the wrapped config.[1](#user-content-fn-1-1fc86e8ada9923692179247ec0e35bf3)
> 
> That's a breaking change, since some users may be relying on system config being ignored.
> 
> IMO neither approach is “better,” and either may be more suitable depending on the situation and the user's intention.
> 
> I could be persuaded either way regarding which approach is better for _most_ users and should be used by the wrapper; or if both techniques were supported, which should be the default behaviour.
> ## Footnotes
> 
>     1. and support for the `exrc` option, but I _was_ oversimplifying [↩](#user-content-fnref-1-1fc86e8ada9923692179247ec0e35bf3)

I dont necessary want to add a "use viminit" option to toggle between -u and VIMINIT just to avoid breaking changes where people were relying on the program not working correctly.

You are right, it is a debate whether --set or --set-default is better. Thats up to the majority opinion on this PR.

MattSturgeon commented:
> just to avoid breaking changes where people were relying on the program not working correctly.

I don't see a "wrap with config" wrapper _ignoring_ system config so that _only_ the wrapped config is used as "not working correctly".

Users who wrap a program with their own custom config typically expect that to be the _only_ config that the program will read; therefore it is usually assumed that config in `/etc`, `XDG_CONFIG_HOME`, and `XDG_CONFIG_DIRS` will be ignored.

Perhaps the issue here is that the same wrapper is used both for wrapping neovim with a custom config _and_ for the usual un-configured `pkgs.neovim` package?

mrcjkb commented:
The current neovim wrapper doesn't ignore rtp scripts (plugin/ftplugin/...) in `$XDG_CONFIG_HOME/$NVIM_APPNAME`, so it isn't pure anyway.

khaneliman commented:
> The current neovim wrapper doesn't ignore rtp scripts (plugin/ftplugin/...) in `$XDG_CONFIG_HOME/$NVIM_APPNAME`, so it isn't pure anyway.

That's fair, would be nice to refactor it a bit to make the impurity more controlled.

BirdeeHub commented:
```nix
{
  inputs = {
    nixpkgs = {
      url = "github:nixos/BirdeeHub/vim_o_exrc_support";
    };
    neovim-nightly = {
      url = "github:nix-community/neovim-nightly-overlay";
      # inputs.nixpkgs.follows = "nixpkgs";
    };
  };
  outputs = { nixpkgs, neovim-nightly, ...}@inputs: let
    overlayMyNeovim = nvim: prev: final: {
      myNeovim = let
        luaRC = final.writeText "init.lua" ''
          require('onedark').setup { style = 'darker', }
          require('onedark').load()
          vim.cmd.colorscheme('onedark')

          vim.o.exrc = true

          require("nvim-treesitter.configs").setup({
              auto_install = false,
              highlight = {
                  enable = true
              },
          })
        '';
      in
      final.wrapNeovim final.neovim-unwrapped {
        configure = {
          customRC = ''lua dofile("${luaRC}")'';
          packages.all.start = with final.vimPlugins; [ 
            nvim-treesitter.withAllGrammars
            onedark-nvim
          ];
          packages.all.opt = with final.vimPlugins; [
          ];
        };
        # extraMakeWrapperArgs = builtins.concatStringsSep " " [
        #   ''--prefix PATH : "${final.lib.makeBinPath (with final; [ stdenv.cc.cc ])}"''
        # ];
        # extraLuaPackages = (_: []);
        # extraPythonPackages = (_: []);
        # withPython3 = true;
        # extraPython3Packages = (_: []);
        # withNodeJs = false;
        # withRuby = true;
        # vimAlias = false;
        # viAlias = false;
        # extraName = "";
      };
    };
    forAllSys = nixpkgs.lib.genAttrs nixpkgs.lib.platforms.all;
  in 
  {
    packages = forAllSys (system: let
      pkgs = import nixpkgs {
        inherit system;
        overlays = [ overlayMyNeovim ];
      };
      pkgs_nightly = import nixpkgs {
        inherit system;
        overlays = [ neovim-nightly.overlays.default overlayMyNeovim ];
      };
    in
    {
      default = pkgs.myNeovim;
      nightly = pkgs_nightly.myNeovim;
    });
  };
}
```

BirdeeHub commented:
for your testing convenience

```nix
{
  inputs = {
    nixpkgs = {
      url = "github:BirdeeHub/nixpkgs/vim_o_exrc_support";
    };
    neovim-nightly = {
      url = "github:nix-community/neovim-nightly-overlay";
      # inputs.nixpkgs.follows = "nixpkgs";
    };
  };
  outputs = { nixpkgs, neovim-nightly, ...}@inputs: let
    overlayMyNeovim = prev: final: {
      myNeovim = let
        luaRC = final.writeText "init.lua" ''
          require('onedark').setup { style = 'darker', }
          require('onedark').load()
          vim.cmd.colorscheme('onedark')

          vim.o.exrc = true

          require("nvim-treesitter.configs").setup({
              auto_install = false,
              highlight = {
                  enable = true
              },
          })
        '';
      in
      final.wrapNeovim final.neovim-unwrapped {
        configure = {
          customRC = ''lua dofile("${luaRC}")'';
          packages.all.start = with final.vimPlugins; [ 
            nvim-treesitter.withAllGrammars
            onedark-nvim
          ];
          packages.all.opt = with final.vimPlugins; [
          ];
        };
        # extraMakeWrapperArgs = builtins.concatStringsSep " " [
        #   ''--prefix PATH : "${final.lib.makeBinPath (with final; [ stdenv.cc.cc ])}"''
        # ];
        # extraLuaPackages = (_: []);
        # extraPythonPackages = (_: []);
        # withPython3 = true;
        # extraPython3Packages = (_: []);
        # withNodeJs = false;
        # withRuby = true;
        # vimAlias = false;
        # viAlias = false;
        # extraName = "";
      };
    };
    forAllSys = nixpkgs.lib.genAttrs nixpkgs.lib.platforms.all;
  in 
  {
    packages = forAllSys (system: let
      pkgs = import nixpkgs {
        inherit system;
        overlays = [ overlayMyNeovim ];
      };
      pkgs_nightly = import nixpkgs {
        inherit system;
        overlays = [ neovim-nightly.overlays.default overlayMyNeovim ];
      };
    in
    {
      default = pkgs.myNeovim;
      nightly = pkgs_nightly.myNeovim;
    });
  };
}
```

BirdeeHub commented:
pros to --set

more secure, can still be overridden inside the wrapper with --set VIMINIT but not from the command line

cons to --set

cant be set from the command line

pros to --set-default

can be overridden from the command line by setting VIMINIT

cons to --set-default

Overriding from the command line is not very useful because you cant, say, run some config and then call the normal config, because the wrapper doesnt tell you the path of the config.

can be overridden from the command line, and that could possibly be considered less safe

k1ng440 commented:
This is PR is useful because I can use `.nvim.lua` to configure the nixd and other LSP for the project. 

```lua
local project_root = vim.fs.root(0, {
    ".git",
})

vim.lsp.config["nixd"] = {
    cmd = { 'nixd' },
    single_file_support = true,
    filetypes = { 'nix' },

    settings = {
        nixpkgs = {
           expr = 'import (builtins.getFlake "' .. project_root .. '").inputs.nixpkgs { }   '
        },
        formatting = {
            command = { "nix", "fmt" },
        },
        options = {
            nixos = {
                expr = '(builtins.getFlake "' .. project_root .. '").nixosConfigurations.xenomorph.options',
            },
            flake_parts = {
                expr = '(builtins.getFlake "' .. project_root .. '").debug.options',
            },
            flake_parts2 = {
                expr = '(builtins.getFlake "' .. project_root .. '").currentSystem.options',
            },
            home_manager = {
                expr = '(builtins.getFlake "' .. project_root .. '").homeConfigurations."k1ng@xenomorph".options',
            },
        },
    },
}

vim.cmd [[
  set makeprg=nix\ flake\ check\ --show-trace
]]
```

BirdeeHub commented:
I found docs that mentioned `-u`

The doc formatting check is failing, but its failing for Typst

mrcjkb commented:
Is there anything blocking this?

@BirdeeHub if the build failure is unrelated, perhaps a rebase will fix it.
Otherwise, we'll have to fix it in another PR.

BirdeeHub committed:
Merge branch 'NixOS:master' into vim_o_exrc_support

BirdeeHub commented:
I am unsure I did that correctly...

But yeah the docs error was for typst

mrcjkb commented on doc/languages-frameworks/neovim.section.md:
```diff
@@ -59,7 +59,7 @@ For instance, `sqlite-lua` needs `g:sqlite_clib_path` to be set to work. Nixpkgs
 - `wrapperArgs`: Extra arguments forwarded to the `makeWrapper` call.
 - `wrapRc`: Nix, not being able to write in your `$HOME`, loads the
-  generated Neovim configuration via its  `-u` argument, i.e. : `-u /nix/store/...generatedInit.lua`. This has side effects like preventing Neovim from reading your config in `$XDG_CONFIG_HOME` (see bullet 7 of [`:help startup`](https://neovim.io/doc/user/starting.html#_initialization) in Neovim). Disable it if you want to generate your own wrapper. You can still reuse while reusing the logic of  ...
+  generated Neovim configuration via the `$VIMINIT` environment variable, i.e. : `export VIMINIT='lua dofile("/nix/store/...generatedInit.lua")'`. This has side effects like preventing Neovim from reading your config in `$XDG_CONFIG_HOME` (see bullet 7 of [`:help startup`](https://neovim.io/doc/user/starting.html#_initialization) in Neovim). Disable it if you want to generate your own wrapper. Yo ...
```
suggestion: `plugin/` scripts in `$XDG_CONFIG_HOME` are still sourced. It's just `init.lua` that isn't, afaict.

```suggestion
  generated Neovim configuration via the `$VIMINIT` environment variable, i.e. : `export VIMINIT='lua dofile("/nix/store/...generatedInit.lua")'`. This has side effects like preventing Neovim from sourcing your `init.lua` in `$XDG_CONFIG_HOME` (see bullet 7 of [`:help startup`](https://neovim.io/doc/user/starting.html#_initialization) in Neovim). Disable it if you want to generate your own wrapper. You can still reuse while reusing the logic of the nixpkgs wrapper and access the generated config via `neovim.passthru.initRc`.
```

BirdeeHub committed:
Update doc/languages-frameworks/neovim.section.md

Co-authored-by: Marc Jakobi <mrcjkb89@outlook.com>

Update pkgs/applications/editors/neovim/wrapper.nix

Co-authored-by: Matt Sturgeon <matt@sturgeon.me.uk>

Update doc/languages-frameworks/neovim.section.md

Co-authored-by: Matt Sturgeon <matt@sturgeon.me.uk>

Update doc/release-notes/rl-2511.section.md

Co-authored-by: Matt Sturgeon <matt@sturgeon.me.uk>

Update doc/languages-frameworks/neovim.section.md

Co-authored-by: Matt Sturgeon <matt@sturgeon.me.uk>

Update doc/release-notes/rl-2511.section.md

Co-authored-by: Matt Sturgeon <matt@sturgeon.me.uk>

Update doc/release-notes/rl-2511.section.md

Co-authored-by: Matt Sturgeon <matt@sturgeon.me.uk>

MattSturgeon reviewed (changes requested):
> Is there anything blocking this?

I was keen to wait for 25.05 to branch off, but that's done now.

However, IMO this is a backwards incompatibility, albeit a subtle one. Therefore it should have a release note in [doc/release-notes/rl-2511.section.md](https://github.com/NixOS/nixpkgs/blob/b72ff2f539d8bded23cbf152ad15189a99306025/doc/release-notes/rl-2511.section.md?plain=1#L12).

Ideally something that explains:

1. what was changed (`-u` -> `VIMINIT`)
2. why it was changed (support `vim.o.exrc`, anything else?)
3. why it is backwards incompatible (less "pure"; more things are sourced from default config locations; maybe list examples, or better yet, a canonical list of things that are now sourced that weren't before?)
4. how users can restore the old behaviour (using `-u` themselves in `wrapperArgs`)

BirdeeHub commented:
> However, IMO this is a backwards incompatibility, albeit a subtle one. Therefore it should have a release note in [doc/release-notes/rl-2511.section.md](https://github.com/NixOS/nixpkgs/blob/b72ff2f539d8bded23cbf152ad15189a99306025/doc/release-notes/rl-2511.section.md?plain=1#L12).

To double check, because I have never heard of this

I just edit that file, correct?

MattSturgeon commented:
> > However, IMO this is a backwards incompatibility, albeit a subtle one. Therefore it should have a release note in [doc/release-notes/rl-2511.section.md](https://github.com/NixOS/nixpkgs/blob/b72ff2f539d8bded23cbf152ad15189a99306025/doc/release-notes/rl-2511.section.md?plain=1#L12).
> 
> To double check, because I have never heard of this
> 
> I just edit that file, correct?

Yes. Since you're the first entry, you'd replace the line saying "put an entry here".

An entry should be a markdown list item. See the 25.05 release notes (rl-2505.section.md) for examples.

You can build and browse the generated docs with `nix-shell ./doc --run devmode`.

> Also, it says
> 
> > To avoid merge conflicts, consider adding your item at an arbitrary place in the list instead.
> 
> I am not 100% sure what this means.

Merge conflicts happen when multiple PRs touch the same line in the same file. They're suggesting that (once the list has grown) you insert your entry in a random spot in the list. If everyone added to the start/end of the list, that'd lead to a lot of merge conflicts.

It doesn't make much sense currently, since there are no entries in the list yet. So if another PR adds an entry, we'll _have_ to deal with conflicts.

BirdeeHub commented:
Well, that explains my confusion then. I was confused because there is no list yet. Makes sense.

BirdeeHub commented:
Ill work on a note in the meantime, maybe I will put it in additions and improvements, im not sure, but ill get some text down and then we can put it wherever

mrcjkb commented:
> I'd be keen to hear where others think the best place for the release note is, and how in-depth it ought to be.

sysinit.vim isn't very widely used (if it were, there would probably be support for sysinit.lua by now).
I think it's highly unlikely to affect anyone on NixOS, but I don't know if that assumption holds for other distros.
Perhaps a short note about the change is a good idea, but I wouldn't bother providing a way to restore the old behaviour unless lots of people ask for it.

BirdeeHub commented:
> > I'd be keen to hear where others think the best place for the release note is, and how in-depth it ought to be.
> 
> sysinit.vim isn't very widely used (if it were, there would probably be support for sysinit.lua by now). I think it's highly unlikely to affect anyone on NixOS, but I don't know if that assumption holds for other distros. Perhaps a short note about the change is a good idea, but I wouldn't bother providing a way to restore the old behaviour unless lots of people ask for it.

I am in agreement with this statement.

I added a release note which I believe to be adequate.

MattSturgeon commented on pkgs/applications/editors/neovim/wrapper.nix:
```diff
@@ -193,8 +193,9 @@ let
-            "-u ${writeText "init.lua" rcContent}"
+            "--set-default"
+            "VIMINIT"
+            "lua dofile([[${writeText "init.lua" rcContent}]])"
```
It should be safe to assume a nix store path won't contain `'`, `"`, _or_ `]` characters, since NAR hashes won't and the filename doesn't.

Would using normal quotes be more readable?

```suggestion
            "lua dofile('${writeText "init.lua" rcContent}')"
```

If we wanted to be safe, we could use `lib.generators.toLua` to escape the string. Internally, `toLua` uses `builtins.toJSON` to serialise strings.

```suggestion
            "lua dofile(${toLua { } (writeText "init.lua" rcContent)})"
```

Or

```suggestion
            "lua dofile(${toJSON (writeText "init.lua" rcContent)})"
```

This could be useful when a non-default nix store is used, as such a store path could potentially contain unexpected characters.

BirdeeHub replied:
Hmmm

I think we should use `'` instead of `[[`

writeText will never give us a path containing `'` but it might give us one containing `]]

more validation than that is unneccessary.

MattSturgeon commented on doc/languages-frameworks/neovim.section.md:
```diff
@@ -59,7 +59,7 @@ For instance, `sqlite-lua` needs `g:sqlite_clib_path` to be set to work. Nixpkgs
 - `wrapperArgs`: Extra arguments forwarded to the `makeWrapper` call.
 - `wrapRc`: Nix, not being able to write in your `$HOME`, loads the
-  generated Neovim configuration via its  `-u` argument, i.e. : `-u /nix/store/...generatedInit.lua`. This has side effects like preventing Neovim from reading your config in `$XDG_CONFIG_HOME` (see bullet 7 of [`:help startup`](https://neovim.io/doc/user/starting.html#_initialization) in Neovim). Disable it if you want to generate your own wrapper. You can still reuse while reusing the logic of  ...
+  generated Neovim configuration via the `$VIMINIT` environment variable, i.e. : `export VIMINIT='lua dofile("/nix/store/...generatedInit.lua")'`. This has side effects like preventing Neovim from sourcing your `init.lua` in `$XDG_CONFIG_HOME/nvim` (see bullet 7 of [`:help startup`](https://neovim.io/doc/user/starting.html#_initialization) in Neovim). Disable it if you want to generate your own w ...
```
Below, you pass `"init.lua"` to the writer, so the actual store path will end in `-init.lua` not `generatedInit.lua`.

```suggestion
  generated Neovim configuration via the `$VIMINIT` environment variable, i.e. : `export VIMINIT='lua dofile("/nix/store/…-init.lua")'`. This has side effects like preventing Neovim from sourcing your `init.lua` in `$XDG_CONFIG_HOME/nvim` (see bullet 7 of [`:help startup`](https://neovim.io/doc/user/starting.html#_initialization) in Neovim). Disable it if you want to generate your own wrapper. You can still reuse while reusing the logic of the nixpkgs wrapper and access the generated config via `neovim.passthru.initRc`.
```

MattSturgeon reviewed (commented):
I've tried to review in a bit more detail, as I think this is more or less ready to be merged.

If any comment feels wrong, out of scope, or too nitpicky feel free to discuss

BirdeeHub commented on pkgs/applications/editors/neovim/wrapper.nix:
```diff
@@ -48,7 +48,7 @@ let
 
       # additional argument not generated by makeNeovimConfig
-      # it will append "-u <customRc>" to the wrapped arguments
+      # it sets VIMINIT variable to "lua dofile(<customRc>)"
```
option 2 but add the quotes.

1 moment I will add this last one and sqash

GaetanLepage commented on doc/languages-frameworks/neovim.section.md:
```diff
@@ -59,7 +59,7 @@ For instance, `sqlite-lua` needs `g:sqlite_clib_path` to be set to work. Nixpkgs
 - `wrapperArgs`: Extra arguments forwarded to the `makeWrapper` call.
 - `wrapRc`: Nix, not being able to write in your `$HOME`, loads the
-  generated Neovim configuration via its  `-u` argument, i.e. : `-u /nix/store/...generatedInit.lua`. This has side effects like preventing Neovim from reading your config in `$XDG_CONFIG_HOME` (see bullet 7 of [`:help startup`](https://neovim.io/doc/user/starting.html#_initialization) in Neovim). Disable it if you want to generate your own wrapper. You can still reuse while reusing the logic of  ...
+  generated Neovim configuration via the `$VIMINIT` environment variable, i.e. : `export VIMINIT='lua dofile("/nix/store/…-init.lua")'`. This has side effects like preventing Neovim from sourcing your `init.lua` in `$XDG_CONFIG_HOME/nvim` (see bullet 7 of [`:help startup`](https://neovim.io/doc/user/starting.html#_initialization) in Neovim). Disable it if you want to generate your own wrapper. Yo ...
```
"You can still reuse while reusing the logic of the nixpkgs wrapper and access the generated config via `neovim.passthru.initRc`."
Isn't this sentence weird?

BirdeeHub replied:
Edit: nvm I stand corrected

BirdeeHub replied:
ugh it is vim with stable and lua with unstable...

BirdeeHub replied:
you misunderstand

The contents.

both contain vimscript

for a moment, I though wrapNeovimUnstable output lua via that value and wrapNeovim output vimscript, but this is not true, both output vimscript

I am aware that both wrapNeovim, that wasn't what was being said.

BirdeeHub commented:
I think that addresses it?

MattSturgeon commented on doc/release-notes/rl-2511.section.md:
```diff
@@ -33,4 +33,6 @@
-- Create the first release note entry in this section!
+- Neovim Wrapper: Added support for the `vim.o.exrc` option, as well as the `VIMINIT` environment variable, and sourcing of `sysinit.vim`.
+
+  See the neovim help page [`:help startup`](https://neovim.io/doc/user/starting.html#startup) for more information, as well as [the nixpkgs neovim wrapper documentation](#neovim-custom-configuration)
```
```suggestion
  See the neovim help page [`:help startup`](https://neovim.io/doc/user/starting.html#startup) for more information, as well as [the nixpkgs neovim wrapper documentation](#neovim-custom-configuration).
```

MattSturgeon commented on doc/languages-frameworks/neovim.section.md:
```diff
@@ -59,7 +59,7 @@ For instance, `sqlite-lua` needs `g:sqlite_clib_path` to be set to work. Nixpkgs
 - `wrapperArgs`: Extra arguments forwarded to the `makeWrapper` call.
 - `wrapRc`: Nix, not being able to write in your `$HOME`, loads the
-  generated Neovim configuration via its  `-u` argument, i.e. : `-u /nix/store/...generatedInit.lua`. This has side effects like preventing Neovim from reading your config in `$XDG_CONFIG_HOME` (see bullet 7 of [`:help startup`](https://neovim.io/doc/user/starting.html#_initialization) in Neovim). Disable it if you want to generate your own wrapper. You can still reuse while reusing the logic of  ...
+  generated Neovim configuration via the `$VIMINIT` environment variable, i.e. : `export VIMINIT='lua dofile("/nix/store/…-init.lua")'`. This has side effects like preventing Neovim from sourcing your `init.lua` in `$XDG_CONFIG_HOME/nvim` (see bullet 7 of [`:help startup`](https://neovim.io/doc/user/starting.html#_initialization) in Neovim). Disable it if you want to generate your own wrapper. Yo ...
```
Let's use the same URL as the release note:

```suggestion
  generated Neovim configuration via the `$VIMINIT` environment variable, i.e. : `export VIMINIT='lua dofile("/nix/store/…-init.lua")'`. This has side effects like preventing Neovim from sourcing your `init.lua` in `$XDG_CONFIG_HOME/nvim` (see bullet 7 of [`:help startup`](https://neovim.io/doc/user/starting.html#startup) in Neovim). Disable it if you want to generate your own wrapper. You can still reuse the generated vimscript init code via `neovim.passthru.initRc`.
```

MattSturgeon commented on doc/release-notes/rl-2511.section.md:
```diff
@@ -33,4 +33,6 @@
 
-- Create the first release note entry in this section!
+- Neovim Wrapper: Added support for the `vim.o.exrc` option, as well as the `VIMINIT` environment variable, and sourcing of `sysinit.vim`.
+
```
Did you check this renders correctly?

Looking at rl2505, it may be more common to format this as a nested list item.

MattSturgeon replied:
LGTM: 
![image](https://github.com/user-attachments/assets/b6fc4380-a9d6-44e2-ba03-268c16414f94)

MattSturgeon commented on doc/release-notes/rl-2511.section.md:
```diff
@@ -33,4 +33,6 @@
 ### Additions and Improvements {#sec-nixpkgs-release-25.11-lib-additions-improvements}
 
-- Create the first release note entry in this section!
+- Neovim Wrapper: Added support for the `vim.o.exrc` option, as well as the `VIMINIT` environment variable, and sourcing of `sysinit.vim`.
```
I'd need to check the rl2505 entries, but I think release notes usually refer to the package attrname:


```suggestion
- `neovim`: Added support for the `vim.o.exrc` option, the `VIMINIT` environment variable, and sourcing of `sysinit.vim`.
```

Or they might use `pkgs.neovim` (I can't remember)

MattSturgeon reviewed (approved):
LGTM. Sorry my reviews were a bit piecemeal!

BirdeeHub commented:
I think you were the only one with any comments left on it tbh. Maybe @GaetanLepage can merge it now?

GaetanLepage commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 395991`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 1 package blacklisted:</summary>
  <ul>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>nixpkgs-manual</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:fast_forward: 1 package blacklisted:</summary>
  <ul>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>nixpkgs-manual</li>
  </ul>
</details>

---
### `x86_64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>nixpkgs-manual</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>nixpkgs-manual</li>
  </ul>
</details>

GaetanLepage reviewed (approved):
Thanks!
